native/membrane_videocompositor_wgpu/src/lib.rs

```diff
@@ -62,41 +58,116 @@ impl InnerState {
 #[rustler::nif]
 fn init(
     #[allow(unused)] env: rustler::Env,
-    upper_caps: RawVideo,
-    lower_caps: RawVideo,
     output_caps: RawVideo,
 ) -> Result<(rustler::Atom, rustler::ResourceArc<State>), rustler::Error> {
     Ok((
         atoms::ok(),
-        rustler::ResourceArc::new(State::new(upper_caps, lower_caps, output_caps)),
+        rustler::ResourceArc::new(State::new(output_caps)),
     ))
 }
 
 #[rustler::nif]
 fn join_frames<'a>(
     env: rustler::Env<'a>,
     state: ResourceArc<State>,
-    upper: rustler::Binary,
-    lower: rustler::Binary,
+    frames: Vec<(usize, rustler::Binary)>,
 ) -> Result<(rustler::Atom, rustler::Term<'a>), rustler::Error> {
     let state = state.lock().unwrap();
 
+    for (i, frame) in &frames {
+        state.compositor.upload_texture(*i, frame)?
+    }
+
     let mut output =
         rustler::OwnedBinary::new(state.output_caps.width * state.output_caps.height * 3 / 2)
             .unwrap(); //FIXME: return an error instead of panicking here
 
-    pollster::block_on(state.compositor.join_frames(
-        upper.as_slice(),
-        lower.as_slice(),
-        output.as_mut_slice(),
-    ));
+    pollster::block_on(state.compositor.draw_into(output.as_mut_slice()));
 
     Ok((atoms::ok(), output.release(env).to_term(env)))
 }
 
+#[rustler::nif]
+fn add_video(
+    #[allow(unused)] env: rustler::Env<'_>,
+    state: rustler::ResourceArc<State>,
+    id: usize,
+    input_video: RawVideo,
+    position: Position,
+) -> Result<rustler::Atom, rustler::Error> {
+    let mut state: std::sync::MutexGuard<InnerState> = state.lock().unwrap();
+    let placement = determine_video_placement(&state, &input_video, &position);
+
+    state
+        .compositor
+        .add_video(id, placement, input_video.width, input_video.height);
+    Ok(atoms::ok())
+}
+
+/// Maps point `x` from the domain \[`x_min`, `x_max`\] to the point in the \[`y_min, y_max`\] line segment, using linear interpolation.
+///
+/// `x` outside the original domain will be extrapolated outside the targe domain.
+fn lerp(x: f64, x_min: f64, x_max: f64, y_min: f64, y_max: f64) -> f64 {
+    (x - x_min) / (x_max - x_min) * (y_max - y_min) + y_min
+}
+
+fn determine_video_placement(
```

Comment:
Maybe this logic should be placed in `native/membrane_videcompositor_wgpu/src/compositor.rs in Scene impl? @DominikWolek